### PR TITLE
DI Refactor and Media Gallery Adjustments

### PR DESCRIPTION
**Pull request by MatkoTom**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant
motivation and context. List any dependencies that are required for this change.

Refactored modules and calling classes which were injected.

We are now using repository classes instead of impl classes. This way we do not depend on impl classes changing their logic.

Also, moved to MutableStateFlow in the MediaFragment.kt and refactored some code there.

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as
  expected)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can
reproduce. Please also list any relevant details for your test configuration. All testing was done
manually by following, plainly, functionality requirements. If there are some special edge cases or
special ways that things need to be tested they will be mentioned below.

Dev testing

**Test Configuration**:

* Firmware version:
* Hardware: SAMSUNG Galaxy S22
* SDK: Android 14

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt [hunk 1]
```diff
@@ -127,10 +129,18 @@ class MediaFragment : BaseFragment() {
                         }
                     }
 
-                    else -> Timber.d("Media error")
+                    is MediaScreenState.Error -> {
+                        // Ignore
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt:133]: Maybe some Toast for error?

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt [hunk 2]
```diff
@@ -152,7 +162,7 @@ class MediaFragment : BaseFragment() {
 
                 message = mediaList[position]
 
-                if (position == smallMediaAdapter?.itemCount) {
+                if (position == smallMediaAdapter?.itemCount?.minus(1)) {
                     viewModel.fetchNextMediaSet(roomId = roomId)
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt:166]: Is this working? If not add TODO

**Review by ivolarevic — DISMISSED** [review]
Generally LGTM but check comments.

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt (hunk 1 above)
Comment (MatkoTom) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt:133]: This is a general state where no media was fetched. Currently, I guess it would only be a black screen? We could think about another implementation here but a Toast could be fine for now.

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt (hunk 2 above)
Comment (MatkoTom) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt:174]: It should be working as intended now.

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatViewModel.kt [hunk 3]
```diff
@@ -283,25 +289,28 @@ class ChatViewModel @Inject constructor(
...
+            Resource.Status.ERROR -> _mediaState.value = MediaScreenState.Error(response.message)
+            Resource.Status.LOADING -> _mediaState.value = MediaScreenState.Loading
+            else -> Timber.d("Other error")
         }
     }
 
     fun fetchNextMediaSet(roomId: Int) {
         val currentLimit = mediaItemsLimit.value ?: 0
-        if (getMediaCount(roomId) > currentLimit) mediaItemsLimit.value = currentLimit + 5
+        if (getMediaCount(roomId) > currentLimit) mediaItemsLimit.value = currentLimit + 50
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatViewModel.kt:311]: Maybe extract this 5 and 50 to const val like MEDIA_LIMIT, MEDIA_OFFSET

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt [hunk 4]
```diff
@@ -111,26 +119,36 @@ class MediaFragment : BaseFragment() {
...
-        viewModel.mediaListener.observe(viewLifecycleOwner,
-            EventObserver { media ->
-                when (media.status) {
-                    Resource.Status.SUCCESS -> {
-                        if (media.responseData != null) {
-                            mediaList = media.responseData
+        lifecycleScope.launch {
+            // Use this below if you need to reload state when app gets back into focus
+//            repeatOnLifecycle(Lifecycle.State.STARTED) {
+            viewModel.mediaState.collect { state ->
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt:125]: do we need state reloading?

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt [hunk 5]
```diff
@@ -152,14 +170,13 @@ class MediaFragment : BaseFragment() {
 
                 message = mediaList[position]
 
-                if (position == smallMediaAdapter?.itemCount) {
+                if (position == smallMediaAdapter?.itemCount?.minus(1)) {
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt:173]: is this final? what changes if you remove .minus(1) like before?

**Review by ivolarevic — APPROVED** [review]
LGTM, but check comments.

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt (hunk 5 above)
Comment (MatkoTom) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt:173]: This change enables the user to continue scrolling when reaching the last item. With the current implementation you will be stopped at the last item, then it will load, and then you will be able to move forward. This change makes for better UX.

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt (hunk 4 above)
Comment (MatkoTom) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/media/MediaFragment.kt:125]: I will enable it. Just because there might be a case where this is beneficial (i.e. fragment goes to the stopped state). Furthermore, we should have no issue with lifecycle data because everything is set before the onStart() method.

**Review by ivolarevic — APPROVED** [review]
LGTM
